### ISSUE TITLE
Add dynamic channel expansion.

### ISSUE THREAD

**Issue by veritas9872** [One-sixth/ms_ssim_pytorch#1]
Hello. I have been looking at the current implementation and I have found that it uses .repeat() on CPU.

However, I would like to propose using .expand() inside the ssim() function. This has several  advantages.

First, this will significantly reduce the amount of data passing from CPU to GPU, which is a major bottleneck in CUDA programming. .repeat() copies data. .expand() does not copy data. So there is not much overhead in using .expand() after transferring the kernel to GPU.

Second, this will allow one to remove the 'channel' parameter from SSIM(), which will remove an unnecessary parameter.

**Comment by One-sixth** [comment]
@veritas9872 Thanks for your advice. I think so too, I just tested the expand operation, but found that performance has decreased. It seems that pytorch jit is not working as we think. This is the modified branch. The time statistics below have been updated.
https://github.com/One-sixth/ms_ssim_pytorch/tree/flexible_multi_channel

**Comment by veritas9872** [comment]
@One-sixth Hello. Thank you for the information on performance. I am glad to know this.
As for the small increase in time, I think that expanding the kernel in ssim() and ms_ssim() functions might reduce the overhead somewhat. The current implementation has the .expand() inside gaussian_kernel(), which is used several times in each function.

**Comment by One-sixth** [comment]
I have modified the code and update branch, but the strange thing is that the time spent does not seem to change. The pytorch jit optimization may be weird. Pytorch jit may have considered this situation and optimized it.

ssim
old
```
Performance Testing SSIM

testing losser2
cuda time 88984.921875
perf_counter time 86.7958464

testing losser3
cuda time 36126.84375
perf_counter time 36.0654904

testing losser4
cuda time 34692.1953125
perf_counter time 33.896039900000005

```
new
```
Performance Testing SSIM

testing losser2
cuda time 88994.9140625
perf_counter time 86.8065547

testing losser3
cuda time 36120.90625
perf_counter time 36.0597451

testing losser4
cuda time 35769.17578125
perf_counter time 35.1033664

```

**Comment by veritas9872** [comment]
@One-sixth Does the new implementation use dynamic value scaling by any chance?
I implemented the value range to be the range of values for the target if the value range was not specified.

**Comment by One-sixth** [comment]
@veritas9872 No, not in my implementation. I think users should set these values manually instead of using automatic.

**Comment by One-sixth** [comment]
I sorted out the code. The new dynamic channel version is available at https://github.com/One-sixth/ms_ssim_pytorch/tree/dynamic_channel_num .

**Comment by veritas9872** [comment]
@One-sixth Thanks for the update!
I have one more question. Do both versions use the 'compensation' variable?
Transferring data from CPU to GPU is actually a very expensive operation and I think that this might be the cause of the difference in speed.

**Comment by One-sixth** [comment]
@veritas9872 Sorry. I don't really understand the meaning of "compensation". My English is not good.
Transferring data from the CPU to the GPU is of course a very expensive operation. But through
```
Mod = SSIM()
Mod = mod.cuda()
```
All parameters will be moved to the GPU. you can run this
```
print(mod.window.device)
```
to found the variable is in the gpu. Because of the use of ScriptModule, all python variables such as self.data_range will become jit constants and merge in jit script. I think my code should be without CPU to GPU performance loss.

**Comment by veritas9872** [comment]
@One-sixth I was referring to the variable `compensation = 1.0` in ssim.py. 
However, I checked and both versions have the variable.

**Comment by One-sixth** [comment]
@veritas9872 I don't know... You need to ask @VainF.

**Comment by VainF** [comment]
The variable comes from [ssim of Tensorflow](https://github.com/tensorflow/tensorflow/blob/r1.14/tensorflow/python/ops/image_ops_impl.py#L3040-L3042)
